Decrypting from stored protocol buffer file fails

After writing a protobuf file and reopening the app, reading the encrypted file fails with the following exception:

``` java
  Caused by: java.io.IOException: No matching key found for the ciphertext in the stream.
                 	at com.google.crypto.tink.streamingaead.InputStreamDecrypter.read(InputStreamDecrypter.java:176)
                 	at com.google.protobuf.CodedInputStream$StreamDecoder.read(CodedInputStream.java:2080)
                 	at com.google.protobuf.CodedInputStream$StreamDecoder.tryRefillBuffer(CodedInputStream.java:2831)
                 	at com.google.protobuf.CodedInputStream$StreamDecoder.isAtEnd(CodedInputStream.java:2754)
                 	at com.google.protobuf.CodedInputStream$StreamDecoder.readTag(CodedInputStream.java:2107)
                 	at com.google.protobuf.CodedInputStreamReader.getFieldNumber(CodedInputStreamReader.java:82)
                 	at com.google.protobuf.MessageSchema.mergeFromHelper(MessageSchema.java:3913)
                 	at com.google.protobuf.MessageSchema.mergeFrom(MessageSchema.java:3895)
                 	at com.google.protobuf.GeneratedMessageLite.parsePartialFrom(GeneratedMessageLite.java:1647)
```

I am using `v1.0.0-alpha03` of the library.

Among other cases this type of error occurs when the outputstream the file is written to is not closed. This can be observed when using the regular `EncryptedFile` and not closing its outpur stream after write. Loading the file again fails with the same exception.

However looking at the datastore implementation... they make use if the `file.use { ... }` function which would indicate a close operation. Feedback is appreciated.

I have recreated a the issue in a minimal snippet: [encrypted-sandbox.zip](https://github.com/osipxd/encrypted-datastore/files/10247056/encrypted-sandbox.zip).

Hope this helps to recreate the issue. Store data. Kill and restart the app:

``` java
Caused by: java.io.IOException: No matching key found for the ciphertext in the stream.
at com.google.crypto.tink.streamingaead.InputStreamDecrypter.read(InputStreamDecrypter.java:176)
```

The snippet serializes to string.

```
object StringSerializer : Serializer<String> {

    override val defaultValue: String = ""

    override suspend fun readFrom(input: InputStream): String {
        try {
            return input.reader().readText()
        } catch (exception: Exception) {
            throw CorruptionException("Can not read string", exception)
        }
    }

    override suspend fun writeTo(t: String, output: OutputStream) =
        output.writer().write(t)

}
```
the datastore
```
        val dataStore = DataStoreFactory.createEncrypted(StringSerializer) {
            EncryptedFile.Builder(
                dataStoreFile("filename"),
                this,
                MasterKeys.getOrCreate(MasterKeys.AES256_GCM_SPEC),
                EncryptedFile.FileEncryptionScheme.AES256_GCM_HKDF_4KB
            ).build()
        }
```
It doesn't get much simpler.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,10 @@
 ## [Unreleased]
 
+### Fixed
+
+- Fixed the case when data can not be read if output stream was not closed in serializer (#10)
+- Added default value for parameter `encryptionOptions` in `PreferenceDataStoreFactory.createEncrypted` (#12)
+
 ## [1.0.0-alpha03] - 2022.11.18
 
 #### More high-level library `security-crypto-datastore`
@@ -73,4 +78,4 @@ To continue use it, change the dependency module in your build script:
 - Migrate dependencies to version catalogs
 
 [unreleased]: https://github.com/osipxd/encrypted-datastore/compare/v1.0.0-alpha03...main
-[v1.0.0-alpha03]: https://github.com/osipxd/encrypted-datastore/compare/v1.0.0-alpha02...v1.0.0-alpha03
+[1.0.0-alpha03]: https://github.com/osipxd/encrypted-datastore/compare/v1.0.0-alpha02...v1.0.0-alpha03

--- a/encrypted-datastore/src/main/kotlin/EncryptingSerializer.kt
+++ b/encrypted-datastore/src/main/kotlin/EncryptingSerializer.kt
@@ -86,7 +86,9 @@ internal class StreamingAeadEncryptingSerializer<T>(
 
     override suspend fun readEncryptedFrom(input: InputStream): T {
         return try {
-            delegate.readFrom(streamingAead.newDecryptingStream(input, associatedData))
+            streamingAead.newDecryptingStream(input, associatedData).use { decryptingStream ->
+                delegate.readFrom(decryptingStream)
+            }
         } catch (e: IOException) {
             throw e.toFriendlyException()
         }
@@ -96,7 +98,8 @@ internal class StreamingAeadEncryptingSerializer<T>(
         return if (isProbablyEncryptedWithAeadException()) {
             CorruptionException(
                 "Can not decrypt DataStore using StreamingAead.\n" +
-                        "Probably you should add decryption fallback to Aead:\n" +
+                        "Probably you have not closed output stream in the `writeTo` method of the serializer or \n" +
+                        "if you had used library before make sure you have added fallback to Aead:\n" +
                         "https://github.com/osipxd/encrypted-datastore#migration",
                 cause = this,
             )
@@ -106,7 +109,9 @@ internal class StreamingAeadEncryptingSerializer<T>(
     }
 
     override suspend fun writeTo(t: T, output: OutputStream) {
-        delegate.writeTo(t, streamingAead.newEncryptingStream(output, associatedData))
+        streamingAead.newEncryptingStream(output, associatedData).use { encryptingStream ->
+            delegate.writeTo(t, encryptingStream)
+        }
     }
 }
 

--- a/encrypted-datastore/src/test/kotlin/EncryptingSerializerTest.kt
+++ b/encrypted-datastore/src/test/kotlin/EncryptingSerializerTest.kt
@@ -1,0 +1,47 @@
+package io.github.osipxd.datastore.encrypted
+
+import androidx.datastore.core.Serializer
+import com.google.crypto.tink.KeyTemplates
+import com.google.crypto.tink.KeysetHandle
+import com.google.crypto.tink.StreamingAead
+import com.google.crypto.tink.streamingaead.StreamingAeadConfig
+import kotlinx.coroutines.runBlocking
+import java.io.ByteArrayOutputStream
+import java.io.InputStream
+import java.io.OutputStream
+import kotlin.test.Test
+import kotlin.test.assertEquals
+
+internal class EncryptingSerializeTest {
+
+    private var streamingAead: StreamingAead
+
+    init {
+        StreamingAeadConfig.register()
+
+        streamingAead = KeysetHandle.generateNew(KeyTemplates.get("AES256_GCM_HKDF_4KB"))
+            .getPrimitive(StreamingAead::class.java)
+    }
+
+    @Test
+    fun `encrypt and decrypt using serializer not closing stream`() {
+        val plaintext = "Plaintext".toByteArray()
+        val encryptingSerializer = AlwaysOpenStreamSerializer.encrypted(streamingAead)
+
+        runBlocking {
+            val bytesOutputStream = ByteArrayOutputStream()
+            encryptingSerializer.writeTo(plaintext, bytesOutputStream)
+
+            val bytes = bytesOutputStream.toByteArray()
+            val decrypted = encryptingSerializer.readFrom(bytes.inputStream())
+
+            assertEquals(plaintext.contentToString(), decrypted.contentToString())
+        }
+    }
+
+    object AlwaysOpenStreamSerializer : Serializer<ByteArray> {
+        override val defaultValue: ByteArray = byteArrayOf()
+        override suspend fun readFrom(input: InputStream): ByteArray = input.readBytes()
+        override suspend fun writeTo(t: ByteArray, output: OutputStream) = output.write(t)
+    }
+}

--- a/gradle/libs.versions.toml
+++ b/gradle/libs.versions.toml
@@ -11,6 +11,8 @@ androidx-datastore-core = { module = "androidx.datastore:datastore-core", versio
 androidx-datastore-preferences = { module = "androidx.datastore:datastore-preferences", version.ref = "datastore" }
 androidx-datastore-preferences-core = { module = "androidx.datastore:datastore-preferences-core", version.ref = "datastore" }
 androidx-security-crypto = "androidx.security:security-crypto:1.0.0"
+androidx-test-core = "androidx.test:core:1.5.0"
+androidx-test-runner = "androidx.test:runner:1.5.1"
 tink = "com.google.crypto.tink:tink-android:1.7.0"
 
 infrastructure-android = { module = "com.redmadrobot.build:infrastructure-android", version.ref = "infrastructure" }

--- a/security-crypto-datastore-preferences/build.gradle.kts
+++ b/security-crypto-datastore-preferences/build.gradle.kts
@@ -7,10 +7,19 @@ description = "AndroidX Security Kotlin Extensions (with DataStore support)"
 
 android {
     namespace = "$group.security.datastore.preferences"
+
+    defaultConfig {
+        testInstrumentationRunner = "androidx.test.runner.AndroidJUnitRunner"
+        testInstrumentationRunnerArguments["clearPackageData"] = "true"
+    }
 }
 
 dependencies {
     implementation(project(":encrypted-datastore-preferences"))
     api(project(":security-crypto-datastore"))
     api(libs.androidx.datastore.preferences)
+
+    androidTestImplementation(kotlin("test", version = libs.versions.kotlin.get()))
+    androidTestImplementation(libs.androidx.test.runner)
+    androidTestImplementation(libs.androidx.test.core)
 }

--- a/security-crypto-datastore-preferences/build.gradle.kts
+++ b/security-crypto-datastore-preferences/build.gradle.kts
@@ -6,8 +6,6 @@ plugins {
 description = "AndroidX Security Kotlin Extensions (with DataStore support)"
 
 android {
-    namespace = "$group.security.datastore.preferences"
-
     defaultConfig {
         testInstrumentationRunner = "androidx.test.runner.AndroidJUnitRunner"
         testInstrumentationRunnerArguments["clearPackageData"] = "true"

--- a/security-crypto-datastore-preferences/src/androidTest/kotlin/EncryptedPreferenceDataStoreTest.kt
+++ b/security-crypto-datastore-preferences/src/androidTest/kotlin/EncryptedPreferenceDataStoreTest.kt
@@ -1,0 +1,59 @@
+package io.github.osipxd.security.crypto
+
+import android.content.Context
+import androidx.datastore.core.DataStore
+import androidx.datastore.dataStoreFile
+import androidx.datastore.preferences.core.PreferenceDataStoreFactory
+import androidx.datastore.preferences.core.Preferences
+import androidx.datastore.preferences.core.edit
+import androidx.datastore.preferences.core.stringPreferencesKey
+import androidx.security.crypto.EncryptedFile
+import androidx.security.crypto.MasterKeys
+import androidx.test.core.app.ApplicationProvider
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.runBlocking
+import org.junit.After
+import org.junit.Test
+import kotlin.test.assertEquals
+
+internal class EncryptedPreferenceDataStoreTest {
+
+    private val context = ApplicationProvider.getApplicationContext<Context>()
+    private val file = context.dataStoreFile("test.preferences_pb")
+
+    @Test
+    fun encryptAndDecrypt(): Unit = runBlocking {
+        val dataKey = stringPreferencesKey("testKey")
+        val plaintext = "Plaintext"
+
+        // Write data and close DataStore
+        coroutineScope {
+            val dataStore = createDataStore(this)
+            dataStore.edit { it[dataKey] = plaintext }
+        }
+
+        // Read data
+        val dataStore = createDataStore(this)
+        val preferences = dataStore.data.first()
+
+        assertEquals(plaintext, preferences[dataKey])
+    }
+
+    @After
+    fun tearDown() {
+        file.delete()
+    }
+
+    private fun createDataStore(scope: CoroutineScope): DataStore<Preferences> {
+        return PreferenceDataStoreFactory.createEncrypted(scope = scope) {
+            EncryptedFile.Builder(
+                file,
+                context,
+                MasterKeys.getOrCreate(MasterKeys.AES256_GCM_SPEC),
+                EncryptedFile.FileEncryptionScheme.AES256_GCM_HKDF_4KB
+            ).build()
+        }
+    }
+}

--- a/security-crypto-datastore-preferences/src/main/AndroidManifest.xml
+++ b/security-crypto-datastore-preferences/src/main/AndroidManifest.xml
@@ -1,0 +1,1 @@
+<manifest package="io.github.osipxd.security.datastore.preferences" />

